static/js/components/PostDisplay.js

```diff
@@ -23,30 +27,41 @@ const postTitle = (post: Post) =>
       {post.title}
     </a>
 
-export default class PostDisplay extends React.Component {
+class PostDisplay extends React.Component {
   props: {
     post: Post,
     showChannelLink?: boolean,
-    expanded?: boolean
+    expanded?: boolean,
+    dispatch: Dispatch
   }
 
   showChannelLink = () => {
     const { post, showChannelLink } = this.props
 
-    return showChannelLink && post.channel_name
+    return (showChannelLink && post.channel_name
       ? <Link to={`/channel/${post.channel_name}`}>
           on {post.channel_name}
       </Link>
-      : null
+      : null)
+  }
+
+  toggleUpvote = () => {
+    const { dispatch, post } = this.props
+    dispatch(actions.postUpvotes.post(post.id, !post.upvoted)).then(post => {
+      dispatch(setPostData(post))
+    })
```

Comment:
I think to me it would make the most sense in this case to stick with the page level container defining the API functions, and then handing them down to the presentational components. If we want to make the `PostDisplay` a container component (as we're doing here with the `connect()` call) I think it would probably make sense to centralize the relevant API functions in that container so that they aren't scattered between e.g. `PostPage` and `PostDisplay`. However, to me the overall pattern that makes the most sense is to have only a few, top-level containers, and for those containers to sort of 'feed' the presentational components data and callbacks. Does that all make sense?

If there's a concern with code duplication we could do something like the following:

```js
toggleUpvote = R.curry(async (dispatch, post) => {
  let post = await dispatch(actions.postUpvotes.post(post.id, !post.upvoted))
  dispatch(setPostData(post))
})
```

then in our container components we would do:

```js
import { toggleUpvote } from 'post-api-utils' // or something
...
const { dispatch } = this.props
return <PostList toggleUpvote={toggleUpvote(dispatch)} />
```